networkx/algorithms/polynomials/tutte.py

```diff
@@ -0,0 +1,180 @@
+    >>> nx.tutte_polynomial(C)
+    x**4 + x**3 + x**2 + x + y
+
+    >>> D = nx.diamond_graph()
+    >>> nx.tutte_polynomial(D)
+    x**3 + 2*x**2 + 2*x*y + x + y**2 + y
+
+    Notes
+    -----
+    This function implements the deletion-contraction recurrence (Definition 3,
+    described above). Edge contraction is defined and deletion-contraction is
+    introduced in [3]_. Combinatorial meaning of the coefficients is introduced
+    in [4]_. Universality, properties, and applications are discussed in [5]_.
+
+    See Also
+    --------
+    _get_cut_edges
+
+    References
+    ----------
+    .. [1] Y. Shi, M. Dehmer, X. Li, I. Gutman,
+       "Graph Polynomials," p. 14
+    .. [2] Y. Shi, M. Dehmer, X. Li, I. Gutman,
+       "Graph Polynomials," p. 46
+    .. [3] D. B. West,
+       "Introduction to Graph Theory," p. 84
+    .. [4] G. Coutinho,
+       "A brief introduction to the Tutte polynomial"
+       Structural Analysis of Complex Networks, 2011
+       https://homepages.dcc.ufmg.br/~gabriel/seminars/coutinho_tuttepolynomial_seminar.pdf
+    .. [5] J. A. Ellis-Monaghan, C. Merino,
+       "Graph polynomials and their applications I: The Tutte polynomial"
+       Structural Analysis of Complex Networks, 2011
+       https://arxiv.org/pdf/0803.3079.pdf
+    """
+    import sympy
+
+    G = nx.MultiGraph(G)
+
+    components = list(nx.connected_components(G))
+    if len(components) > 1:
+        t = 1
+        for component in components:
+            t *= tutte_polynomial(G.subgraph(component), False)
+        return sympy.simplify(t)
+
+    cut_edges = _get_cut_edges(G)
+    loops = [e for e in G.edges if e[0] == e[1]]
+    edges_not_cuts_loops = [i for i in G.edges if i not in cut_edges and i not in loops]
+
+    if not edges_not_cuts_loops:
+        return sympy.Symbol("x") ** len(cut_edges) * sympy.Symbol("y") ** len(loops)
+
+    e = edges_not_cuts_loops[0]
+    # deletion-contraction
+    C = nx.contracted_edge(G, e, self_loops=True)
+    contraction_loops = [i for i in C.edges if i[0] == i[1] and i not in G.edges]
+    if contraction_loops:
+        C.remove_edge(*contraction_loops[0])
+    G.remove_edge(*e)
+
+    t = tutte_polynomial(G, False) + tutte_polynomial(C, False)
+    if simplify:
+        return sympy.simplify(t)
+    return t
+
+
+def _get_cut_edges(G):
+    """Finds the cut-edges of a provided graph.
+
+    Cut-edges are edges whose deletion increases the number of components of a
+    graph [1]_. A cut-edge is an edge that is not a member of any cycle.
+
+    Parameters
+    ----------
+    G : NetworkX graph
+
+    Returns
+    -------
+    list
+        A list of all cut-edges of `G`.
+
+    References
+    ----------
+    .. [1] D. B. West,
+       "Introduction to Graph Theory," p. 23
+    """
+    cut_edges = []
+    for e in G.edges:
+        G_copy = G.copy()
+        G_copy.remove_edge(*e)
+        if len(list(nx.connected_components(G_copy))) > 1:
+            cut_edges.append(e)
```

Comment:
This looks really really slow. Are you really going to copy the graph for every edge in the graph?
Also, you don't need to form the list of connected components if all you want is the number of them.
Try `nx.number_connected_components` instead.

Finally, why is your thinking behind making this a private function?

Reply:
Good point. I could just remove the edge and add it back, instead, e.g.:

```python
G_copy = G.copy()
cut_edges = []
for e in G.edges:
    G_copy.remove_edge(*e)
    if len(list(nx.connected_components(G_copy))) > 1:
        cut_edges.append(e)
    G_copy.add_edge(*e)
```

I can see why this should be a public function. Do you think it would make more sense to include in cuts module?